tests/Highlighter.py

```diff
@@ -0,0 +1,358 @@
+    "blue"                : "S",
+    "S"                   : "S",
+
+    "invalid"             : "I",
+    "inv"                 : "I",
+    "pink"                : "I",
+    "I"                   : "I",
+}
+
+
+class HighlightTester(unittest.TestCase):
+
+    def assert_highlighted_chr(self, code, expected, level, lang="en"):
+        rules  = self.get_rules(level, lang)
+        self.check(code, expected, rules)
+
+
+    def assert_highlighted_chr_multi_line(self, *args, level, lang="en"):
+        if len(args)%2 != 0:
+            raise RuntimeError(f'Pass an even number of strings to assert_highlighted_chr_multi_line (alternatingly code and highlighting). Got: {args}')
+
+        code = []
+        expected = []
+        for i, line in enumerate(args):
+            if i%2 == 0: code.append(line)
+            else: expected.append(line)
+
+        code = "\n".join(code)
+        expected = "\n".join(expected)
+
+        self.assert_highlighted_chr(code, expected, level, lang)
+
+
+    def assert_highlighted(self, code_col, level, lang="en"):
+        code, expected = self.converts(code_col)
+        self.assert_highlighted_chr(code, expected, level, lang)
+
+
+    def assert_highlighted_multi_line(self, *args, level, lang="en"):
+        code_col = "\n".join(args)
+        code, expected = self.converts(code_col)
+        self.assert_highlighted_chr(code, expected, level, lang)
+
+
+
+    # apply rules on code and check if the result is valid
+    def check(self, code, expected, rules):
+        
+        simulator = SimulatorAce(rules)
+        result = simulator.highlight(code)
+
+        valid, ind_error = self.compare(result, expected)
+
+        if not valid:
+
+            code_list     = code.split('\n')
+            expected_list = expected.split('\n')
+            result_list   = result.split('\n')
+
+            line_cpt = 0
+            while ind_error >= len(code_list[line_cpt]):
+                ind_error -= len(code_list[line_cpt]) +1
+                line_cpt += 1
+
+            print("ERROR in this code :")
+            print("In this code :", code_list[line_cpt] )
+            print("We want      :", expected_list[line_cpt] )
+            print("We have      :", result_list[line_cpt] )
+            print("At           :", " " * ind_error + "^")
+
+        self.assertTrue(valid)
+
+
+    # This function allows to compare 2 syntactic colorations one desired
+    # and the other obtained, and manages some special cases
+
+    # Returns a boolean and a value:
+    # - If the 2 highlights are consistent, the boolean will be true and the value unused
+    # - if there is an error in the highlight, the boolean will be false and
+    #   the value will be the number of the 1st character where there is an inconsistency
+    def compare(self, result, expected):
+        if len(result) != len(expected):
+            raise ValueError("The desired highlight and the obtained highlight do not have the same length !")
```

Comment:
Here's another thing to thing about for fitness of purpose:

If we hit this error case, there will be no useful output for a developer to go on. They will only see the error message you wrote above. They will not see the expected code, the actual code, the highlights, not the lengths, nothing.

That's not setting debuggers up to succeed, right?

Reply:
> And about the code display, it would be convenient, but I don't have access in the context of this function.

Well. You *could*, since you wrote this function. You can add arguments to also pass the code in, right? 😊

What I'm saying is this:

* First, think about the **user** this code is for. What task are they trying to perform? What would help them with that task? What should they see?
* Then, with that experience in mind, write the code to pass the right information to the right place.

Looks like you did it the other way around: 

* First, write some code that does some computational task
* Then, see what kind of message you can scramble together from the variables you happen to have access to based on the code written in step 1

A lot of people work in the second style. The problem with that style is it works great *if and only if* everything works out. But when things go wrong, it's usually very hard to figure out *where*, and *why*. 

Plan for things to go wrong, always, all the time, and design your software accordingly.

Reply:
This function does not exist anymore. Since the checks are done in the check method, I have access to the code variable (and since I use the assertEqual method, I don't have to write the error message)